sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedMicroBatchStream.scala

```diff
@@ -0,0 +1,196 @@
+   * </p>
+   * <p>
+   * This method will be called multiple times, to launch one Spark job for each micro-batch in this
+   * data stream.
+   * </p>
+   */
+  override def planInputPartitions(startOffset: Offset, endOffset: Offset): Array[InputPartition] = {
+    assertNotNull(startOffset, "startOffset")
+    assertNotNull(endOffset, "endOffset")
+    assert(startOffset.isInstanceOf[ChangeFeedOffset], "Argument 'startOffset' is not a change feed offset.")
+    assert(endOffset.isInstanceOf[ChangeFeedOffset], "Argument 'endOffset' is not a change feed offset.")
+
+    val start = startOffset.asInstanceOf[ChangeFeedOffset]
+    val end = endOffset.asInstanceOf[ChangeFeedOffset]
+
+    assert(end.inputPartitions.isDefined, "Argument 'endOffset.inputPartitions' must not be null or empty.")
+
+    end
+      .inputPartitions
+      .get
+      .map(partition => partition
+        .withContinuationState(
+          SparkBridgeImplementationInternal
+            .extractChangeFeedStateForRange(start.changeFeedState, partition.feedRange),
+        false))
+  }
+
+  /**
+   * Returns a factory to create a `PartitionReader` for each `InputPartition`.
+   */
+  override def createReaderFactory(): PartitionReaderFactory = {
+    ChangeFeedScanPartitionReaderFactory(config, schema, cosmosClientStateHandle)
+  }
+
+  /**
+   * Returns the most recent offset available given a read limit. The start offset can be used
+   * to figure out how much new data should be read given the limit. Users should implement this
+   * method instead of latestOffset for a MicroBatchStream or getOffset for Source.
+   *
+   * When this method is called on a `Source`, the source can return `null` if there is no
+   * data to process. In addition, for the very first micro-batch, the `startOffset` will be
+   * null as well.
+   *
+   * When this method is called on a MicroBatchStream, the `startOffset` will be `initialOffset`
+   * for the very first micro-batch. The source can return `null` if there is no data to process.
+   */
+  // This method is doing all the heavy lifting - after calculating the latest offset
+  // all information necessary to plan partitions is available - so we plan partitions here and
+  // serialize them in the end offset returned to avoid any IO calls for the actual partitioning
+  override def latestOffset(startOffset: Offset, readLimit: ReadLimit): Offset = {
+    val startChangeFeedOffset = startOffset.asInstanceOf[ChangeFeedOffset]
+    val offset = CosmosPartitionPlanner.getLatestOffset(
+      startChangeFeedOffset,
+      readLimit,
+      Duration.ZERO,
+      this.clientConfiguration,
+      this.cosmosClientStateHandle,
+      this.containerConfig,
+      this.partitioningConfig,
+      this.session
+    )
+
+    if (offset.changeFeedState != startChangeFeedOffset.changeFeedState) {
+      this.latestOffsetSnapshot = Some(offset)
+      offset
+    } else {
+      // scalastyle:off null
+      this.latestOffsetSnapshot = null
+      null
```

Comment:
Is null the way to signal a non-existing Offset? Isn't it weird that they went with that contract?

Reply:
null means no more data to process - happens because the API interfaces are actually written in Java.

Reply:
Added a comment with explanation in next iteration although the method comment also explains it.